File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashMap<Character, Integer> indexMap = new HashMap<>();
        int l = 0, r = 0;
        int longest = 0;
        while (r < s.length()) {
            char ch = s.charAt(r);
            if (indexMap.containsKey(ch) && indexMap.get(ch) >= l) {
                l = indexMap.get(ch) + 1;
                indexMap.put(ch, r);
            } else {
                indexMap.put(ch, r);
            }
            longest = Math.max(longest, r - l + 1);
            r++;
        }

        return longest;
    }
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        # p w w k e w
        #       i
        #           j
        freq = {}
        i, j = 0, 0
        maxLen = 0
        while j < len(s):
            if s[j] in freq and freq[s[j]] >= i:
                i = freq[s[j]] + 1
            else:
                freq[s[j]] = j
                maxLen = max(maxLen, j - i + 1)
                j += 1

        return maxLen
            
                